feat(AST): Rule based expression translator

fpacifici opened the pull request:
This introduces an expression translator infrastructure we could use to translate Snuba Queries into Clickhouse queries that supports mapping rules defined for each storage and associated to the relationship between entity and storage.

Remark: the files are displayed here in an order that is the opposite to the one that makes understanding easier. I suggest to review them in this order:
- snuba/datasets/plans/translator/visitor.py (all the documentation is here)
- snuba/clickhouse/translator/visitor.py 
- snuba/clickhouse/translator/rules.py

This PR introduces the Expression Translator, which is capable of translating one AST expression into another one, potentially in a different AST.
The translation is executed by scanning the expression tree through a visitor and applying a list of rules for each node.
If a rule applies it returns the translated value, if it does not it returns None.
A followup PR will wire it up into a Query Translator that will translate all the expression in a Query building a new Query.

Multiple sets of rules can be composed to translate joins and (with small extensions) to run a multi step translation for multi table storages (Snuba ast to Clickhouse ast and a second step from Clickhouse ast to Clickhouse ast).

fpacifici committed:
Move translator file

fpacifici committed:
Add translator

fpacifici committed:
Add Cliockhouse translator

fpacifici committed:
Add Clickhouse translator

fpacifici committed:
Add test cases

fpacifici committed:
Remove the matcher

fpacifici committed:
Various re-namings

fpacifici committed:
Allow default values for TranslationRules

fpacifici committed:
Cleanup

fpacifici committed:
Add documentation

fpacifici committed:
Cleanup and bug fix

fpacifici commented on snuba/clickhouse/query.py:
```diff
@@ -1,8 +1,13 @@
+
+# This defines the type of a Clickhouse query expression. It is used to make
+# the interface of the expression translator more clear.
+Expression = NewType("Expression", SnubaExpression)
```
This really exists only to make this easier to understand so that we know exactly what depends on the logical expression and what depends on the clickhouse expression.

fpacifici commented on snuba/clickhouse/translator/rules.py:
```diff
@@ -0,0 +1,85 @@
+
+
+@dataclass(frozen=True)
+class ColumnMapper(ExpressionMapper[Column, Expression]):
```
These two rules (column and tags) are just example to show how this will work. 
See tests/clickhouse/translator/test_translation.py for examples of how to connect everything.

If/when we will introduce more complex expressions in the snuba language (that will require a more complex translation) we could introduce a matcher based ExpressionMapper, that is configured with a matcher that matches complex expression and finds parameters (like a regex) and provides the parameters to an executor that builds the result expression

fpacifici commented on snuba/clickhouse/translator/visitor.py:
```diff
@@ -0,0 +1,125 @@
+TSimpleExp = TypeVar("TSimpleExp", bound=SnubaExpression)
+
+
+class DefaultSimpleMapper(ExpressionMapper[TSimpleExp, ClickhouseExpression]):
```
These (from here on) are all the defaults that just produce a copy of the expression. For each structured expression, the children are delegated back to the visitor so, for each node, we can apply the appropriate mapper.

fpacifici commented on snuba/datasets/plans/translator/visitor.py:
```diff
@@ -0,0 +1,158 @@
+    Multi-step (still TODO) translations can be supported by applying a second sequence of
+    rules to the result of the first one for each node in the expression to be translated.
+
+    The visitor approach, while a bit verbose, allows us to impose some static
+    guarantees of correctness of the translator. It is impossible, for example, to add a
+    new node type to the Snuba AST and ignoring it in the translator (a method
+    implementation would be missing).
```
To me the advantage of building this with a visitor is that mapping rules (ExpressionMapper) can be Generic with the expression type as parameter. This mean we do not need to check the expression type with an `isinstance` call at the very beginning of each mapper.

fpacifici commented on snuba/datasets/plans/translator/visitor.py:
```diff
@@ -0,0 +1,158 @@
+
+
+@dataclass(frozen=True)
+class TranslationRules(Generic[TExpOut]):
```
I wanted to put the MappingExpressionTranslator class at the beginning of this file. It would make this whole a lot more clear. The problem is that, while python is fine with that because of __futures__.annotations, mypy is not ok with that and it is not able to resolve the input types passed to MappingExpressionTranslator. __map_expression

lynnagara commented on snuba/clickhouse/translator/rules.py:
```diff
@@ -0,0 +1,85 @@
+@dataclass(frozen=True)
+class TagMapper(ExpressionMapper[SubscriptableReference, Expression]):
+    """
+    Basic implementation of a tag mapper that transforms a subscriptable
```
Does "subscriptable" relate only to nested columns? Or should ColumnMapper get passed a SubscriptableReference like TagMapper?

fpacifici replied:
SubscriptableReference is only loosely connected to the concept of nested column. It means an expression that can be accessed as a subscriptable (like [] in pyrthon). Today this is only for some very specific nested columns (tags) but it may be a simple not nested array in the future.
So yes, today it only relates to some specific nested columns, but it may become more generic.
Remember an expression in the AST is a vehicle to access a database concept (like a column), it does not represent the database concept itself.
So while a nested column is a column in the database, a SubscriptableReference in the ast is not a reference to a Column (thus it is not a Column), it is just an expression to access a value in a mapping column. Does this make sense ?

ColumnMapper cannot be passed a SubscriptableReference, it takes a different type in. This is intentional, the two classes are siblings.

lynnagara replied:
Yes, thanks

fpacifici committed:
Merge branch 'master' into feat/add_translator

fpacifici commented on snuba/datasets/plans/translator/visitor.py:
```diff
@@ -0,0 +1,158 @@
+    applied in a single pass over the AST.
+    This allows us to support joins and multi-step translations (for multi table
+    storages) as an example:
+    Joins can be supported by simply concatenating rule sets associated with each storage.
```
I think it may be useful to give some context on how I envision joins to work since it may not be obvious:
**phase 1:** pre-entity. Here not all columns in the AST will be fully qualified. 
- non join query: Here the columns now are not fully qualified. It works as it is, columns do not reference a table alias and the rules do not care about table aliases
- join query: Here the columns are fully qualified so that they are not ambiguous. Since the translation rules do not consider table aliases we need to do soemthing about it. The Qoery Plan Builder for joins will wrap all the translation rules before applying them to account for the table alias. Something like:
```
# JoinWrapper takes in the column to translate, removes the alias, execute the inner rule
# and transform the output by adding the alias back.
storage1_rules = [JoinWrapper(rule, table_alias="event") for rule in storage1_rules]
storage2_rules = [JoinWrapper(rule, table_alias="groups") for rule in storage2_rules]
rules = [*storage1_rules, *storage2_rules]
translator = Translator(rules)
```

**phase 2:** entities.
Now we don't care about all that complexity anymore since all columns in the AST (and, as a consequences all translation rules) are fully qualified with the entity name, which replaces the table aliases in joins.

fpacifici commented on snuba/clickhouse/translator/rules.py:
```diff
@@ -0,0 +1,85 @@
+    to_col_name: str
+    to_table_name: Optional[str]
+
+    def attemptMap(
```
We discussed in the past whether tag promotion should be in the translator or in a query processor. There were advantages and drawbacks of both solutions.
The advantage of doing it in a processor is that the translator would be simpler and easier to express in a declarative way.
The advantage of doing it in the translator, is that it would be easier to identify tags expressions. This is the example of doing that in a processor: https://github.com/getsentry/snuba/pull/955/files#diff-17bcbc40c66ec928ca0b4e13622325e5R44-R59

I think that verbosity would be mitigated a lot by https://github.com/getsentry/snuba/pull/942 so I don't see a strong advantage in doing tag promotion in the translator which would then need a lot more context on the schema to do tag promotion.

fpacifici commented on snuba/datasets/plans/translator/visitor.py:
```diff
@@ -0,0 +1,158 @@
+        Translates an expression if this rule matches such expression. If not, it
+        returns None.
+
+        It receives an instance of a translating visitor to take care of children
+        that this rule is not capable, or not supposed to, translate. These node
+        should be delegated to the children_translator.
```
Some additional context on why we need to pass the translator to an ExpressionMapper and delegate to the mapper the responsibility to apply the translator to its children:

in the original AST transform method, the FunctionCall node guides this process: it first applies the transformation lambda to all children (parameters), then it replaces the old parameter with the new one on the FunctionaCall node and then it applies the lambda to the new node and returns the result.

I originally wanted to do something like this where the visitor ensures first we translate the parameter of a function then we find a matcher for the function itself with the translated parameters.
Unfortunately the translation produces a different data type than the one it took in input (they are not Expressions since they are the equivalent of the clickhouse query) so that simple process breaks. We cannot pack the logical pre-translation FunctionCall with the post-translation parameters and find a mapper for that. 
Thus we need the mapper to return a fully translated subtree (that type checks properly), and we provide a translator to the mapper so it does not have to figure out how to translate its children.

tkaemming commented on snuba/clickhouse/translator/visitor.py:
```diff
@@ -0,0 +1,125 @@
+        expression: TSimpleExp,
+        children_translator: ExpressionVisitor[ClickhouseExpression],
+    ) -> Optional[ClickhouseExpression]:
+        return ClickhouseExpression(deepcopy(expression))
```
Something isn't adding up for me here.

[`ClickhouseExpression` is a new type derived from `snuba.query.expressions.Expression`](https://github.com/getsentry/snuba/blob/2b5cc5aea0b09fb647a61d176a7b95cfb8b994cc/snuba/clickhouse/query.py#L11-L13) (presumably because it will be very similar but represents a different domain so it makes sense to keep the class lineages separate) — but [`Expression` is an abstract base class](https://github.com/getsentry/snuba/blob/2b5cc5aea0b09fb647a61d176a7b95cfb8b994cc/snuba/query/expressions.py#L19-L64) that only has an `alias` constructor parameter, so instantiating `ClickhouseExpression` should throw, right? That is what `Expression` does when instantiated:

```
>>> from snuba.query.expressions import Expression
>>> Expression('asdf')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class Expression with abstract methods __iter__, accept, transform
```

It seems like at runtime the type constructor becomes the identity function of the input (as if it were a cast), which explains why the tests are passing:

```python
>>> from snuba.clickhouse.translator.visitor import ClickhouseExpression
>>> o = object()
>>> ClickhouseExpression(o) is o
True
```

The runtime behavior is pretty clearly documented [in the `NewType` docs](https://docs.python.org/3/library/typing.html#newtype) — but it also says "The static type checker will treat the new type as if it were a subclass of the original type" which seems to be _not_ consistent with the abstract base class behavior.

For example, if this patch is applied to change `NewType` to a type alias instead (maintaining the _actual_ behavior of the constructor, which is something I'd expect that we want as soon as there is a real `ClickhouseExpression` type), like so:

```python
diff --git a/snuba/clickhouse/query.py b/snuba/clickhouse/query.py
index ab79432..c37034a 100644
--- a/snuba/clickhouse/query.py
+++ b/snuba/clickhouse/query.py
@@ -10,4 +10,4 @@ Query = NewType("Query", LogicalQuery)
 
 # This defines the type of a Clickhouse query expression. It is used to make
 # the interface of the expression translator more clear.
-Expression = NewType("Expression", SnubaExpression)
+Expression = SnubaExpression
```

Then mypy goes bonkers and everything in `tests/clickhouse/translator/test_visitor.py` fails with `TypeError: Can't instantiate abstract class Expression with abstract methods __iter__, accept, transform` (which is what I expected it to do in the first place):

```
snuba/clickhouse/translator/visitor.py:57: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:57: error: Argument 1 to "Expression" has incompatible type "TSimpleExp"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:66: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:67: error: Argument 1 to "Expression" has incompatible type "FunctionCall"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:84: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:85: error: Argument 1 to "Expression" has incompatible type "CurriedFunctionCall"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:105: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:106: error: Argument 1 to "Expression" has incompatible type "SubscriptableReference"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:120: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:121: error: Argument 1 to "Expression" has incompatible type "Lambda"; expected "Optional[str]"
```

Is this something that you were aware of an intended to do? It seems like all of these would have to change to some extent as soon as an actual `ClickhouseExpression` type hierarchy exists — which I think will mostly affect this `DefaultSimpleMapper` since the output types will have to be the concrete types and not the abstract `ClickhouseExpression`.

Luckily, I do think that using `ClickhouseExpression` as the `TExpOut` for `ExpressionMapper` type generally makes sense (one exception in `DefaultCurriedFunctionMapper` noted below.)

fpacifici replied:
`Is this something that you were aware of and intended to do?`
Mostly. To me it is still an open point whether we will have a separate Expression hierarchy between Logical and Clickhouse. We have a reason now in the SubscriptableReference (which does not make sense in clickhouse) but that's it. So there are two options and I followed this approach. (declaring a type for clickhouseExpression) here to avoid getting stuck into a solution that would make it hard to introduce the clickhosue hierarchy.

1) We will have a separate hierarchy. This will have different types inside to an extent
In this case I don;t think there will be a common parent class that is not something above the already abstract ClickhouseExpression and SnubaExpression. In such case translation would not be a deepcopy. So the problem does not exist. We would be instantiating an object of a different type with its own constructor.

2) It will turn out we use the same expression hierarchy between logical and clickhouse
In this case I would argue we should not even have a type alias or a NewType. The translator will take an Expression and produce an Expression, since the hierarchy would be agnostic to whether it is part of the logical or the physical query. 
Again this has a problem in the SubscriptableReference, but, as we discussed today, there is an option in turning it into a FunctionCall. The question is how likely is it that we will be able to keep everything in one Expression hierarchy if/when we start doing some more advanced logical query processing.

3) ok, they were three. The two hierarchies look just the same but we want to keep different naming
I am not sure this would be better than 2 if there is no structural difference between the hierarchies, but in this case I would go with a plain type alias without trying to instantiate an abstract class.

I created the NewType here just to be sure we do not make implementation decisions that prevent us from picking 1 or 2 and leave us the time to figure out whether we actually want a separate Expression hierarchy for Clickhouse. I think it may be likely depending on how often we end up wanting new nodes in the Logical query to do logical query processing.
Anyway we cannot create a separate hierarchy as long as the Query class is still the same.

tkaemming commented on snuba/clickhouse/translator/visitor.py:
```diff
@@ -0,0 +1,125 @@
+        return ClickhouseExpression(
+            replace(
+                expression,
+                internal_function=expression.internal_function.accept(
+                    children_translator
+                ),
```
Won't `accept` return a `ClickhouseExpression` which can't be assigned to `CurriedFunctionCall.internal_function` which would expect a `FunctionCall` (or `ClickhouseFunctionCall`, eventually?) Is this getting lost in `replace`?

fpacifici replied:
Actually no, Adding a new node type to achieve a strict type checking of CurriedFUnctionCall.inner_function won't work. I cannot have (in a visitor) different visit method with different return types (so that visitInnerType would return FunctionCall instead of Expression) since the return type of visitXXX is the same of the `accept` method which is defined in the abstract class Expression.

I tried a different approach which is sketched here (which is a commit in this PR that I reverted so you can still see the current version of the pr):
https://github.com/getsentry/snuba/pull/936/commits/df9097534e6b2d98c1725441419db6b2655d36c2

This is what it does:
1) TranslationRules contains also some dedicated some dedicated expression mappers to translator things like CurriedFunction.inner_function. `ExpressionMapper[FunctionCall, FunctionCall]`
2) The MappingExpressionTranslator extends the Visitor adding also a few methods to translate specific types of expressions so that they can return the type we want. Like `def translate_inner_function(self, exp: FunctionCall) -> TInnerFunctionOut`. Now TInnerFunctionOut does not need to be a generic Expression but it can be specific (FunctionCall).
3) DefaultCurriedFunctionMapper uses this dedicated method to translate internal_function. 

This guarantees strict type checking. I removed replace and it checks. The downside is the amount of type parameters I have to add. Now the ExpressionMapper takes 5 arguments. I am a bit skeptic about the complexity otherwise I would be ok cleaning that part up and bringing that commit back.
What do you think ?

fpacifici replied:
Please disregard the last comment, see this instead 
https://github.com/getsentry/snuba/pull/936#issuecomment-630702483

(leaving it here to document the previous attempt)

tkaemming reviewed (commented):
I've looked through `snuba/datasets/plans/translator/visitor.py` and `snuba/clickhouse/translator/visitor.py` so far. The big picture here makes sense so far, but see inline notes about the details of the ClickHouse translator. Going to come back to the rest a little later.

fpacifici committed:
Merge branch 'master' into feat/add_translator

fpacifici committed:
Rename attemptMap

fpacifici committed:
Merge branch 'feat/add_translator' of https://github.com/getsentry/snuba into feat/add_translator

fpacifici committed:
Test strict typing for Inner Function

fpacifici committed:
Revert "Test strict typing for Inner Function"

This reverts commit df9097534e6b2d98c1725441419db6b2655d36c2.

fpacifici committed:
Make strict type checking work, for real

fpacifici commented:
Took a different approach to deal with the AST nodes that have children that are not bare Expression (like CurriedFunctionCall and SubscriptableReference):
1) Instead of doing static type checking on the input of ExpressionMapper, I made it strict on the output. This means TranslationRules do not have anymore one attribute per input type but one attribute per output type, which allows us to have rules in the same translator which return a different output type (like FunctionCall for CurriedFunctionCall.inner_function).
2) The translator does not implement Visitor but still follows the same idea. It does not have a different method per expression data type it receives but it has a method per data type it has to produce (so it has a translate_inner_function)

This has a few advantages:
- we can actually have proper strict type checking that works 
- we can converge to this same api for both Expression.transform and the visitor used for formatting. This can support all three use cases.

It has some drawbacks:
- verbosity: now we have really a lot of type arguments around (the ExpressionMapper has 6 parameters, though implementation should never see more than 2).

fpacifici committed:
Move responsibility to clickhouse side

fpacifici committed:
Split snuba, rules and clickhouse

fpacifici committed:
Move files around

fpacifici committed:
Merge branch 'master' into feat/add_translator

fpacifici committed:
Make rules more strict

fpacifici committed:
Bring visitor back

fpacifici committed:
Move rules applier

fpacifici committed:
Reorganize files

fpacifici committed:
A better naming

fpacifici committed:
Init Multistep

fpacifici committed:
Add test

fpacifici committed:
Cleanup

fpacifici committed:
Documentation

fpacifici commented:
I tried a different approach to deal with the CurriedFunctionCall issue (the inner_function child is of FunctionCall type so it cannot be translated by a method that blindly return Expression).

Thinking over this problem I changed my mind with respect to the first implementation and I believe it deserved some effort to make it hard/impossible to someone who creates translation rules to cause a type error there.
This is because it would be hard to ensure all rules are compatible (without relying on the type checker) especially because rules do not know anything of their parents so they do not know they are translating the internal function of a CurriedFunctionCall.
Also, while there is only one case of this issue in the AST today, if we split it between snuba and clickhouse the two can evolve independently making it harder to contain the problem.

The approach I followed was to restrict the valid rules for translation in terms of input type and output type. The first implementation provided a Mapper that would take any valid snuba expression and returned potentially any valid clickhouse expression. Any translation rule that extended this (ExpressionMapper) could be used by the translator.
Now there is a specific set of abstract classes that extend ExpressionMapper that defines which combinations of types are allowed (for example they enforce that a function can only be turned into another function). These are declared in allowed.py and the rule based translator can only be initialzied with rules that extend those in allowed.py.
Thus it is impossible to instantiate a translator that tries to do something too broad.

The rest is in line with what I did before, just changed the class organization.

fpacifici commented on snuba/clickhouse/query.py:
```diff
@@ -1,8 +1,13 @@
+
+# This defines the type of a Clickhouse query expression. It is used to make
+# the interface of the expression translator more clear.
+Expression = SnubaExpression
```
I turned this into a simple type alias because, in order build translation rules that accept very specific types (like Column) instead of a generic expression, such types need to be subtypes of Clickhouse Expression.
```
ClickhouseExpression = NewType("ClickhouseExpression", SnubaExpression)
# Now ClickhosueExpression does not have any subclass. And Column, 
# FunctionCall etc. are not ClickhouseExpression making it not very useful.
```

I don't think this is a problem since, creating this type is only meant to chow what depends on a clickhouse expression in the code and what depends on Snuba Expressions for clarity.

There are two options to consolidate this:
If we built a separate clickhouse AST, they will include different classes, they won't be aliases or NewTypes of each other. Likely they won't even have a common parent class. SnubaExpression and CLickhouseExpression will be siblings at most.

If we decide there is no reason to change the Expression AST (the Query class will definitively be different but in this case the expression AST may be the same data structure) then this alias won't be needed at all. We will just use the same object.

fpacifici commented on snuba/clickhouse/translators/snuba/allowed.py:
```diff
@@ -0,0 +1,87 @@
+    pass
+
+
+# Below are the abstract classes that need to be extended in order to write a mapping rule
```
I have been quite arbitrarily restrictive on this file. A lot more than we actually need. Likely we will release some of these constraint when we start writing real translation rules. But I suspect it is not going to change much.

fpacifici commented on snuba/clickhouse/translators/snuba/allowed.py:
```diff
@@ -0,0 +1,87 @@
+    pass
+
+
+class SubscriptableReferenceMapper(
```
If we do not allow SubscriptableReference to be transformed into another  SubscriptableReference we won;t be able to wire up the translator as long as we do not have all the appropriate translation rules in place. Or translation would fail.

fpacifici committed:
Remove useless deepcopy

fpacifici committed:
Merge branch 'master' into feat/add_translator

fpacifici committed:
Make docstring shorter

fpacifici committed:
Remove multistep translator to make this more digestible

fpacifici commented on snuba/clickhouse/translators/snuba/mappers.py:
```diff
@@ -0,0 +1,79 @@
+
+
+@dataclass(frozen=True)
+class SimpleColumnMapper(ColumnMapper):
```
These two rules (column and tags) are just example to show how this will work.
See tests/clickhouse/translator/test_translation.py for examples of how to connect everything.

If/when we will introduce more complex expressions in the snuba language (that will require a more complex translation) we could introduce a matcher based ExpressionMapper, that is configured with a matcher that matches complex expression and finds parameters (like a regex) and provides the parameters to an executor that builds the result expression

fpacifici commented on snuba/clickhouse/translators/snuba/mapping.py:
```diff
@@ -0,0 +1,152 @@
+    This relies on a visitor so that we can statically enforce that no
+    expression subtype is added to the code base without properly support
+    it in all translators. The downside is verbosity.
+    """
```
I think it may be useful to give some context on how I envision joins to work since it may not be obvious:
phase 1: pre-entity. Here not all columns in the AST will be fully qualified.

non join query: Here the columns now are not fully qualified. It works as it is, columns do not reference a table alias and the rules do not care about table aliases
join query: Here the columns in the ast are fully qualified so that they are not ambiguous. Since the translation rules do not consider table aliases we need to do soemthing about it. The Query Plan Builder for joins will wrap all the translation rules before applying them to account for the table alias. Something like:
```
storage1_rules = [JoinWrapper(rule, table_alias="event") for rule in storage1_rules]
storage2_rules = [JoinWrapper(rule, table_alias="groups") for rule in storage2_rules]
# JoinWrapper takes in the column to translate, removes the alias, execute the inner rule
# and transform the output by adding the alias back.

rules = [*storage1_rules, *storage2_rules]
translator = Translator(rules)
```

phase 2: entities or after #966 
Now we don't care about all that complexity anymore since all columns in the AST (and, as a consequences all translation rules) are fully qualified with the entity name, which replaces the table aliases in joins.

tkaemming replied:
This is clever.

fpacifici commented on snuba/datasets/plans/translator/mapper.py:
```diff
@@ -0,0 +1,50 @@
+
+    An ExpressionMapper is supposed to know how to map an expression but
+    it is not required to map its children. For that it receives a
+    translator of TTranslator type.
```
Some additional context on why we need to pass the translator to an ExpressionMapper and delegate to the mapper the responsibility to apply the translator to its children:

in the original AST transform method, the FunctionCall node guides this process: it first applies the transformation lambda to all children (parameters), then it replaces the old parameter with the new one on the FunctionaCall node and then it applies the lambda to the new node and returns the result.

I originally wanted to do something like this where the visitor ensures that first we translate the parameter of a function then we find a matcher for the function itself with the translated parameters.
Unfortunately the translation produces a different data type than the one it takes in input (they are not Expressions since they are the equivalent of the clickhouse query) so that simple process breaks. We cannot pack the logical pre-translation FunctionCall with the post-translation parameters and find a mapper for that because it would not be a valid expression.
Thus we need the mapper to return a fully translated subtree (that type checks properly), and we provide a translator to the mapper so it does not have to figure out how to translate its children.

fpacifici commented on snuba/clickhouse/translators/snuba/__init__.py:
```diff
@@ -0,0 +1,45 @@
+    """
+    Type strict Snuba to Clickhouse translator.
+
+    When translating an expression from an AST to another one we cannot
+    assume that any expression type of the first AST can be translated
+    into any type of expression in the second AST in any context.
```
ok thanks, will rephrase.

fpacifici commented on snuba/clickhouse/translators/snuba/__init__.py:
```diff
@@ -0,0 +1,45 @@
+    translates the inner function is forced to produce a FunctionCall
+    or the AST will be broken.
+
+    The exact constraints on which translations are valid depend on the
+    two ASTs. In the Snuba to Clickhouse translation there is one constraint
+    only (the CurriedFunctionCall outlined above).
```
I did not write this down, but I am forming a pretty solid opinion that, yes, we should revisit CurriedFunctionCall. Specifically I think these two should happen:
1) It should not have a simple FunctionCall as internal function. That is not "any" function. It  only works for some specific functions. From the type structure perspective the CurriedInnerFunction would just look like FunctionCall, but having a different type would simplify a few things here. Or the inner function could be not a function at all.
2) I think CurriedFunctionCall should only be in the clickhouse AST to begin with. I think we are complicating our life for nothing by allowing it to exist in the logical AST. What do you think ? That would mean we would not have to translate it.

tkaemming replied:
1. I agree in general. I think that in retrospect, there's not much reuse value in having the inner function be a `FunctionCall` other than the superficial resemblance in structure. At this point, my guess is that it'd be simpler to replace it with  a `ParametricFunctionCall` that more closely follows [the ClickHouse definition](https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/parametric-functions/) with two different of argument lists. I was curious if `Sequence[Expression]` (or `Tuple`, whatever) would make sense as the type of the first argument list (the "parameters" in their vocabulary), so I just tried the example below out of curiosity, and … well, the implication here will be pretty self-explanatory when you read it, I think. Their design makes lot more sense to me now.

```
SELECT quantile(dummy)(dummy)
FROM system.one
GROUP BY dummy

Received exception from server (version 19.11.14):
Code: 134. DB::Exception: Received from localhost:9000. DB::Exception: Parameters to aggregate functions must be literals. 

0 rows in set. Elapsed: 0.106 sec. 
```

2. I don't have a strong opinion about whether or not it exists in the logical AST, other than my general hesitation around reverse incompatible changes. I guess I haven't felt any real pain from dealing with it at the logical level. If we removed it from the logical query, trying to deal with translating plain function calls back into parametric calls and enforcing the literal requirement seems like some degree of unnecessary complexity (e.g. if `quantile(x)(y)` becomes `quantile(x, y)`), unless we disallow them completely unless statically defined in query processors (basically the equivalent of what Sentry does with percentiles as `pXX(col)` which I don't care for either.)

fpacifici commented on snuba/clickhouse/translators/snuba/allowed.py:
```diff
@@ -0,0 +1,92 @@
+    Functions are only allowed to become Functions so that we can ensure
+    CurriedFunctions internal functions can be successfully translated.
+
+    TODO: We actually need to loosen this constraint (for functions
+    translating into pre aggregated Column). Though before doing so we
+    need to give a dedicated type that will be stricter for the
+    CurriedFunction internal function.
```
FunctionCallMapper here only allows to translate a function into a function.
Any translation we may want to do for a multi table storage (where one of the tables is actually pre-aggregated) would require us to translate FunctionCall into Column. Since we will receive a function (count()) and we will have to translate into a column that has the aggregation.

If I allow this here, that would ruin all the effort to keep translation type safe for CurreidFunctionCall since the inner function cannot be a Column. Though as soon as I swap the inner function with its own type (see comment above) this will not be a problem anymore.

tkaemming replied:
I just saw this after looking at https://github.com/getsentry/snuba/pull/936#discussion_r430861407. This makes more sense now.

fpacifici commented on snuba/clickhouse/translators/snuba/mappers.py:
```diff
@@ -0,0 +1,79 @@
+    The alias is not transformed.
+    """
+
+    from_col_name: str
+    from_table_name: Optional[str]
+    to_col_name: str
+    to_table_name: Optional[str]
```
> Reading this really makes me wish that the alias was separate from the Column type

We did that in the first iteration of the AST, early in the review. If I remember ther reason why we reverted to alias being an attribute was that all expression can have one in clickhouse. 
I think something like this would be a viable change.
```
class AliasedExpression(Expression):
    alias: str
    expression: Expression
```
It would make it harder to mess up though being aliases expressions themselves, they would be still present in the AST and thus translatable and transformable.

Though I am not sure I get what the connection of the rest of the comment is with aliases.

tkaemming replied:
> Though I am not sure I get what the connection of the rest of the comment is with aliases.

Yeah, sorry — it was a pretty bad description and kind of intertwined two somewhat related trains of thought. There were two points here:

1. It seems that there would be two instance attributes in this class in an ideal scenario — `to_column` and `from_column` (ignoring reserved words), both `Columns`, and that the `if` in `attempt_map` would be better expressed as `if expression == self.from_column`. Since the `Column` has the alias on it already, this doesn't really work.
2. If the `path` attribute gets added to `Column` (and no changes to aliasing), it seems like that logic should get also implemented here for consistency — or it might imply that the path is something separate from the `Column` itself, if that equality relationship above doesn't make sense, broadly speaking.

I think I'd be okay with (prefer, really) only allowing aliases in the `SELECT` clause at this point, though I'm not sure if there's anything that happens already that depends on aliasing being available elsewhere (I hope not.)

fpacifici replied:
> I think I'd be okay with (prefer, really) only allowing aliases in the SELECT clause at this point, though I'm not sure if there's anything that happens already that depends on aliasing being available elsewhere (I hope not.)

Guess what? yes there is :( 
And guess what? It is related to tags_key/tags_value. (from very early on, cannot say when this happened).
What happens is that the code that generates the old style arrayJoin operation,  does the join only once with keys and value, produces a bidimensional array, adds an alias to that and then refer to that alias to fetch one column or the other depending if the user wants keys or values 
https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L309
So internally we do use them to avoid (supposedly) calculating twice the same expensive operation.

I don't think the fact that we restricted aliases to the select expression should be a limiting factor to revisit this. I think what we should gauge carefully for the impact on query processing. The downside here is that this mapper should be in sync with the column structure, the downside of having aliases being nodes of the ast themselves means they would have to be taken into account during query processing (like always checking if the expression is a column or a column nested in an alias). This can be mitigated with the pattern system. 
I think it is something to explore.

About this specific mapper, I meant it more as a showcase to illustrate how the translator works. I think there are several possible alternatives on how to map columns. Ideally I would like this to look like more a dictionary where you provide references to the columns from a schema and they are mapped into columns of another schema without caring about each field of the column.

fpacifici commented on snuba/clickhouse/translators/snuba/mappers.py:
```diff
@@ -0,0 +1,79 @@
+    The alias is not transformed.
+    """
+
+    from_col_name: str
+    from_table_name: Optional[str]
+    to_col_name: str
+    to_table_name: Optional[str]
```
indeed. I promised evan that I will let him commit #959 before changing columns for everyone since his PR has been there for longer. As soon as he merges I will adapt 963 if this will be already merged or vice versa.

fpacifici commented on snuba/query/dsl.py:
```diff
@@ -10,6 +10,13 @@ def literals_tuple(alias: Optional[str], literals: Sequence[Literal]) -> Functio
 
 
+# Array functions
+def array_element(
```
you sure about this?  Assuming we fix the visitor, this would probably be the only place in the code base where we do camel case.
Anyway, very weak opinion, I don't think either way ruins readability.

tkaemming replied:
I thought I responded to this but apparently I forgot. I don't have a strong opinion about this, but I think it's OK for the DSL to bend some of the rules if it helps it maintain consistency with the domain it's supposed to represent.

Another example is I think operator overloading is almost always a bad idea, but it seems like it might make sense to replace (or supplement) things like the arithmetic operators with operators on `Expression` that return a new `Expression` with the operator function applied.

fpacifici commented on snuba/query/expressions.py:
```diff
@@ -163,9 +163,7 @@ def accept(self, visitor: ExpressionVisitor[TVisited]) -> TVisited:
-            self,
-            column=self.column.transform(func),
-            key=self.key.transform(func),
+            self, column=self.column.transform(func), key=self.key.transform(func),
```
do you mean reformatting the file? This happened because I had made a change in an earlier version and black decided to reformat. Since it was only one statement I did not bother reverting it.

tkaemming replied:
Yeah, but mostly how/why these changes that were okay at one point aren't now (presuming they were and didn't just slip in.)

fpacifici reviewed (commented):
thanks for the review. A few answers inline

fpacifici committed:
Merge branch 'master' into feat/add_translator

fpacifici committed:
Apply PR comments
